Add new actor window queries

guilherme-brandao opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description

## Link(s) to Ticket(s) or Issue(s) resolved by this PR

## Are these changes tested and documented?

- [ ] If tested, please describe how. If not, why tests are not needed.
- [ ] If documented, please describe where. If not, describe why docs are not needed.
- [ ] Added to `Unreleased` section of `CHANGELOG.md`?

## Still Left Todo

*Fill this out if this is a Draft PR so others can help.*

guilherme-brandao committed:
Added new actor window queries

xmariachi commented on x/emissions/keeper/queryserver/query_server_topics.go:
```diff
@@ -164,3 +165,181 @@ func (qs queryServer) GetNextChurningBlockByTopicId(ctx context.Context, req *ty
+	}
+
+	// Calculate next window timing based on topic epoch
+	nextChurningBlock, isActive, err := qs.k.GetNextPossibleChurningBlockByTopicId(ctx, req.TopicId)
```
Just a note: I first thought that we may also include an `is_active` flag, however then I noticed that this is covered by having a `Next...` values informed (which is the way we check topic activity in `isTopicActive` ), so it would be a bit redundant.  If there is no "next" , then it means also that the topic is inactive... but may not be crystal clear to the user.

wdyt? Should we add `is_active` explicitly for a full picture?

guilherme-brandao replied:
I think it's a good idea to add `isTopicActive` property, but it would be just for workers, right?

xmariachi replied:
Why?
I think it could be for both. 
Currently our way of determining whether a topic is active, is to check whether this topic has a next churning block, right?
But topic activity is "the same" for both workers and reputers.

xmariachi replied:
using for both. Adding linter fixes [here](https://github.com/allora-network/allora-chain/pull/856)

xmariachi commented on x/emissions/keeper/queryserver/query_server_topics.go:
```diff
@@ -164,3 +165,181 @@ func (qs queryServer) GetNextChurningBlockByTopicId(ctx context.Context, req *ty
+				response.CurrentNonceBlockHeight = nonce.ReputerNonce.BlockHeight
+				response.WindowStartBlock = windowStart
+				response.WindowEndBlock = windowEnd
+			}
```
once we found the nonce that applies, we can exit the loop once we find the nonce for efficiency, right?

guilherme-brandao replied:
in this case, we should go until the end to find the most recent one.

xmariachi replied:
I think they are sorted the other way: most recent first.
However a bit below, I make a comment about not just getting the first one - also the next so we can use it for the `next`.

```
 allorad q emissions  unfulfilled-reputer-nonces 63  --node https://allora-rpc.testnet.allora.network   
nonces:
  nonces:
  - reputer_nonce:
      block_height: "5469384"
  - reputer_nonce:
      block_height: "5469324"
  - reputer_nonce:
      block_height: "5469264"
  - reputer_nonce:
      block_height: "5469204"
...
```

xmariachi commented on x/emissions/keeper/queryserver/query_server_topics.go:
```diff
@@ -164,3 +165,181 @@ func (qs queryServer) GetNextChurningBlockByTopicId(ctx context.Context, req *ty
+		}
+	}
+
+	// For next window timing, we need to check unfulfilled worker nonces
```
Nope - They open on next nonce + GroundTruthLag. 
~I think in practice the most economical way to get this, is just `windowStart + EpochLength`, and no reference to worker nonces. 
This assumes topic is active - and still, this is an estimation (topic may be inactivated between now and next window), so it's fine.~
We can get it from the loop before: first iteration get the nonce, second iteration get the next-nonce, then exit - that would be the real info right there).

(Note for current solution as is, that worker nonces do not remain there - so if gtlag is big it won't work , note that there would be a big decalage between current worker nonce, and the projected reputer nonce)

xmariachi reviewed (commented):
Partial review

xmariachi commented on x/emissions/module/autocli.go:
```diff
@@ -673,30 +673,6 @@ func (am AppModule) AutoCLIOptions() *autocliv1.ModuleOptions {
-						{ProtoField: "topic_id"},
-					},
-				},
 				// add inclusion methods
```
The new queries should also be added into the CLI ( `GetReputerSubmissionWindowStatus`, and worker )

guilherme-brandao replied:
added!

xmariachi reviewed (changes requested):
Second leg of the review

guilherme-brandao committed:
Added reviewer suggestion

guilherme-brandao committed:
Add auto cli commands

xmariachi commented on x/emissions/keeper/queryserver/query_server_topics.go:
```diff
@@ -164,3 +165,183 @@ func (qs queryServer) GetNextChurningBlockByTopicId(ctx context.Context, req *ty
+	var earliestFutureNonce *types.ReputerRequestNonce
+
+	for _, nonce := range nonces.Nonces {
+		extraLag := topic.GroundTruthLag % topic.EpochLength
```
these four lines of the extraLag calculation can go outside the loop - they are not dependent on the nonce but on the topic.

xmariachi replied:
refactored [here](https://github.com/allora-network/allora-chain/pull/856)

guilherme-brandao committed:
Add isTopicActive in the window query

xmariachi committed:
add changelog

xmariachi committed:
extraLag redundant, explcit errors, linter

guilherme-brandao committed:
Diego/add actor window queries pr fixes (#856)

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a
completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description
* Always error on `err` 
* extraLag duplicated computations
* linter, `is_active` given a default of false.

## Are these changes tested and documented?

- [ ] If tested, please describe how. If not, why tests are not needed.
- [ ] If documented, please describe where. If not, describe why docs
are not needed.
- [ ] Added to `Unreleased` section of `CHANGELOG.md`?

guilherme-brandao committed:
Improve query

amimart commented on x/emissions/proto/emissions/v9/query.proto:
```diff
@@ -555,6 +540,18 @@ service QueryService {
+    option (google.api.http).get = "/emissions/v9/worker_submission_window_status/{topic_id}";
+  }
+
+  // Get reputer submission window status and timing
```
We can have multiple nonce open at the same time so I think we should query the window status for a nonce, not only a topic.

Also maybe we could have a single query to provide both worker and reputer window information
